### PR TITLE
Update complete-firebase-authentication-react-tutorial.md

### PR DESCRIPTION
**Pull request by EmmanuelStalling**

### CODE REVIEW

File: complete-firebase-authentication-react-tutorial.md [hunk 1]
```diff
@@ -1,365 +1,353 @@
...
-
-* [React's Context API](https://www.robinwieruch.de/react-context-api/)
-* [A gentle introduction to React's Higher Order Components](https://www.robinwieruch.de/gentle-introduction-higher-order-components/)
+The requirements for this tutorial are a working [editor or IDE, a running command line](https://www.robinwieruch.de/developer-setup/), and recent versions of {{% a_blank "node and npm" "https://nodejs.org/en/" %}}. You should have learned about React in the first place. [The Road to learn React](https://www.robinwieruch.de/the-road-to-learn-react/) is a free ebook that provides all the fundamentals of React. You will build a larger application in plain React, and transition from JavaScript ES5 to JavaScript ES6 and beyond. This tutorial will not dive into all the details taught in the ebook, so take the chance to grab your copy of it to learn those first.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "editor or IDE, a running command line" -> "editor or an IDE/terminal"

File: complete-firebase-authentication-react-tutorial.md [hunk 2]
```diff
@@ -1,365 +1,353 @@
...
-  * Bonus: [Authentication in React, Firebase and Redux](#react-firebase-authentication-redux)
-  * Bonus: [Authentication in React, Firebase and MobX](#react-firebase-authentication-mobx)
 
 {{% chapter_header "React Application Setup: create-react-app" "react-application-setup" %}}
 
-You are going to implement a whole authentication mechanism in React with sign up, sign in and sign out. Furthermore, it should be possible to reset a password or change a password as a user. The latter option is only available for an authenticated user. Last but not least, it should be possible to protect certain routes (URLs) to be accessible by authenticated users only. Therefore you will build a proper authorization solution around it.
+There are two ways to begin the lesson for this application: either follow my guidance in this section; or find a starter project in this {{% a_blank "GitHub repository" "hhttps://github.com/the-road-to-react-with-firebase/react-firebase-authentication-starter-kit" %}} and follow its installation instructions. [Editors Note: Tell them why they should choose one or the other, else you're suggesting an alternative route for no reason]
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: Good point:

"[Editors Note: Tell them why they should choose one or the other, else you're suggesting an alternative route for no reason]" -> "Whereas the former will show you how to set up the whole project from scratch, the latter will give you instant access to a starter project where you don't need to worry about setting up the folder/file structure yourself."

File: complete-firebase-authentication-react-tutorial.md [hunk 3]
```diff
@@ -1,365 +1,353 @@
...
+Now we'll set up the project for our needs. First, get rid of the files from the bootstrapped React project, since we won't be using them. From the command line, head to your *src/* folder and execute it:
 
 {{< highlight javascript >}}
-npm install firebase react-router-dom
+cd src
+rm App.js App.test.js App.css logo.svg
 {{< /highlight >}}
 
-In addition, create a *components/* folder in your application's *src/* folder on the command. That's where all your components will be implemented eventually.
+Second, create a *components/* folder in your application's *src/* folder on the command line. This is where all your components will be implemented. Also,- the App component that you have removed in the previous step will be recreated here.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "Also,-" -> "Also,"

File: complete-firebase-authentication-react-tutorial.md [hunk 4]
```diff
@@ -1,365 +1,353 @@
...
+Create a dedicated folder for each component we will implement for this application :
+
+{{< highlight javascript >}}
+cd components
+mkdir Account Admin App Home Landing SignIn SignOut SignUp
+mkdir Navigation PasswordChange PasswordForget
+mkdir Session Firebase
+{{< /highlight >}}
+
+In each folder, create an *index.js* file for the the component. Navigate into a folder, create the file, and navigate out again. Repeat these steps for every component. You can choose to name your folders/files differently, but that's how I liked to do it for my applications.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "the the" -> "the"

File: complete-firebase-authentication-react-tutorial.md [hunk 5]
```diff
@@ -1,365 +1,353 @@
...
+* Optionally, introduce [Git and keep track of your changes by having your project on GitHub](https://www.robinwieruch.de/git-essential-commands/)
+
+{{% chapter_header "React Router for Firebase Auth" "react-router-setup" %}}
+
+Since we are building a larger application in the following sections, it would be great to have a couple of pages (e.g. landing page, account page, admin page, sign up page, sign in page) to split the application into multiple URLs (e.g. /landing, /account, /admin). These URLs or subpaths of a domain are called routes in a client-side web application. Let's implement the routing with {{% a_blank "React Router" "https://github.com/ReactTraining/react-router" %}} before we dive into Firebase for the realtime database and authentication/authorization. 
+
+The application should have multiple routes. For instance, a user should be able to visit a landing page, and also use sign up and sign in pages to enter the application as an authenticated user. If a user is authenticated, it is possible to visit protected pages like account or admin pages.You can consolidate all the routes of your application in a well-defined *src/constants/routes.js* constants file:
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "admin pages.You can consolidate" -> "admin pages. You can consolidate"

File: complete-firebase-authentication-react-tutorial.md [hunk 6]
```diff
@@ -1,365 +1,353 @@
...
 
-Once your project is created on their website, you should have a dashboard for it. There you can find a menu item which says "Authentication". Select it and click "Sign-In Method" menu item afterward. There you can enable the authentication with Email/Password. This tutorial will cover the basic authentication with email and password, but feel free to add other authentication methods later on. If you do so and open source your project, you can link it in the comments below for other readers of this guide.
+The main focus here is using Firebase in React for the application we'll build together. Firebase, bought by Google in 2014, enables realtime databases in React, which can be used for extensive authentication and authorization, and even for deployingment. You can build real-world applications with React and Firebase without worrying about implementing a backend application. All the things a backend application would handle, like authentication and a database, is handled by Firebase. Many businesses use React and Firebase to power their applications, as it is the ultimate combination to launch an {{% a_blank "MVP" "https://en.wikipedia.org/wiki/Minimum_viable_product" %}}.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "deployingment" -> "deployment"

File: complete-firebase-authentication-react-tutorial.md [hunk 7]
```diff
@@ -372,209 +360,385 @@ const config = {
...
+);
 
-// Sign In
-export const doSignInWithEmailAndPassword = (email, password) =>
-  auth.signInWithEmailAndPassword(email, password);
+export default SomeComponent;
 {{< /highlight >}}
 
-At this point, note that these endpoints are called asynchronously. They need to be resolved later on. In addition, there needs to be error handling for it. For instance, it is not possible to sign in a user who is not signed up yet. The Firebase API would return an error. You will implement all of this in a later part of this tutorial.
+Firebase and React are now connected,  the fundamental step to make the layers communicate with one other. Next, we will implement the interface for the Firebase class on our side to communicate with the Firebase API.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "with one other" <- correct English?

File: complete-firebase-authentication-react-tutorial.md [hunk 8]
```diff
@@ -372,209 +360,385 @@ const config = {
...
+}
 
-// Sign out
-export const doSignOut = () =>
-  auth.signOut();
+export default Firebase;
 {{< /highlight >}}
 
-Last but not least, the two optional functions to reset and change a password of an authenticated user:
+Let's define all the authentication functions step by step. They will serve our communication channel from the Firebase class to the Firebase API. First, the sign up function (registration) takes email and password parameters for its function signature and uses an official Firebase API endpoint to create a user.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "authentication functions step by step." -> "authentication functions as class methods step by step."

File: complete-firebase-authentication-react-tutorial.md [hunk 9]
```diff
@@ -372,209 +360,385 @@ const config = {
...
+  doPasswordReset = email => this.auth.sendPasswordResetEmail(email);
+
+  doPasswordUpdate = password =>
+    this.auth.currentUser.updatePassword(password);
+}
+
+export default Firebase;
+{{< /highlight >}}
+
+That's the authentication interface for your React components connected to the Firebase API. It covers all the use cases for this application. In the next section, we will consume all the methods of your Firebase class in your React components.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "for your React components connected to the Firebase API" -> "for your React components eventually connected to the Firebase API"

File: complete-firebase-authentication-react-tutorial.md [hunk 10]
```diff
@@ -372,209 +360,385 @@ const config = {
...
+{{% chapter_header "Sign Up with React and Firebase" "react-firebase-sign-up" %}}
 
-Let's start with the sign up page. It consists of the page, a form, and a link. Whereas the form is used to sign up a new user to your application, the link will be used later on the sign in page when a user has no account yet. It is only a redirect to the sign up page, but not used on the sign up page itself. Still, it shares the same domain and therefore shares the same file as the sign up page and sign up form.
+We set up all the routes for your application, configured Firebase and implemented the authentication API for your Firebase class. Now it's time to use the authentication functionalities in your React components, which we'll build from scratch. I try to put most of the code in one block, because the components are not too small, and splitting them up step by step might be too verbose. Nevertheless, I will guide you through each code block afterward. The code blocks for forms can become repetitive, so they will be explained.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "so they will be explained." -> "so they will be explained here once in an elaborate way."

File: complete-firebase-authentication-react-tutorial.md [hunk 11]
```diff
@@ -834,227 +1057,230 @@ class SignUpForm extends Component {
...
+* Confirm your {{% a_blank "source code for the last section" "https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/commit/6cbccce21952e10472504fc6ef3d37af2c998bb6" %}}.
 
-In order to complete the authentication loop, let's implement as last part the sign out component. It doesn't need any form and is only a button which shows up in the Navigation component in the next steps. Since we can use the already defined authentication API to sign out a user, it is fairly straight forward to pass the functionality to a button in a React component.
+{{% chapter_header "Sign Out with React and Firebase" "react-firebase-sign-out" %}}
 
-In *src/components/SignOut.js* file:
+To complete the authentication loop, next we'll implement the sign out component. The component is just a button that appears within the Navigation component. Since we can use the previously-defined authentication API to sign out a user, passing functionality to a button in a React component is fairly straightforward. Implement the SignOut in the *src/components/SignOut/index.js* file:
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "previously-defined" -> "previously defined"

File: complete-firebase-authentication-react-tutorial.md [hunk 12]
```diff
@@ -1114,118 +1384,144 @@ class App extends Component {
...
+* Confirm your {{% a_blank "source code for the last section" "https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/tree/8fa558878fa5c7e7bbb905c0c12b592cbfd69e3d" %}}.
 
-First, you can revert the recent changes in the App component. It can become a functional stateless component again. It doesn't need to know about the authenticated user object anymore.
+{{% chapter_header "Session Handling with Higher-Order Components" "react-firebase-session-handling-higher-order-components" %}}
 
-In *src/components/App.js* file:
+We added a basic version of session handling in the last section. However,  the authenticated user still needs to be passed down from the App component to interested parties. That ca becomes tedious over time, because the authenticated user has to be passed through all components until it reaches all the component in the leaf components. You used the React Context API to pass down the Firebase instance to any component before. Here, you will do the same for the authenticated user. In a new *src/components/Session/context.js* file, place the following new React Context for the session (authenticated user):
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "That ca becomes" -> "That becomes"

Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "it reaches all the component in the leaf components" -> "it reaches all the leaf components"

File: complete-firebase-authentication-react-tutorial.md [hunk 13]
```diff
@@ -1114,118 +1384,144 @@ class App extends Component {
...
-  </Router>
+{{< highlight javascript >}}
+import AuthUserContext from './context';
 
-export default App;
+export { AuthUserContext };
 {{< /highlight >}}
 
-Next, you can wrap the App component up in a session handling higher order component. That's where all the business logic goes which you have just removed in the last step from the App component. Basically you enhance the App component with the higher order component.
+The App component can use the new context to provide the authenticated user to omponents that are interested in it:
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: omponents -> components

File: complete-firebase-authentication-react-tutorial.md [hunk 14]
```diff
@@ -1492,1143 +1759,673 @@ class PasswordChangeForm extends Component {
...
+* Research yourself how a role-based or permission-based authorization could be implemented
+* Confirm your {{% a_blank "source code for the last section" "https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/tree/94bb3a92970d4159564eb01bcfd24bcc0f8c4348" %}}
 
-First of all, create a file for your Firebase realtime database API. It goes into the firebase folder next to your file for the authentication API.
+{{% chapter_header "Firebase Realtime Database in React" "react-firebase-realtime-database" %}}
 
-From *src/firebase/* folder:
+So far, only Firebase knows about your users. There is no way to retrieve a single user or a list of users. They are stored internally by Firebase to keep the authentication secure. That's good, because you are never involved in storing sensible data like passwords. However, you can introduce the Firebase realtime database to keep track of user entitiesm otherwise you can't associate other domain entities (e.g. a message, book, invoice) created by your users to your users. This section will explain how to store users in your realtime database in Firebase.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: entitiesm -> entities

File: complete-firebase-authentication-react-tutorial.md [hunk 15]
```diff
@@ -1492,1143 +1759,673 @@ class PasswordChangeForm extends Component {
...
+}
+
+export default Firebase;
 {{< /highlight >}}
 
-Finally, you can use those functionalities in your React components to create and retrieve users. Let's start with the user creation. The best place to do it would be the SignUpForm component. It is the most natural place to create a user in the database after the sign up through the Firebase authentication API. You can simply add another API request to create a user when the user signed up successfully.
+The paths in the `ref()` method match the location where your entities (users) will be stored in Firebase's realtime API. If you delete a user at "users/5", the user with the identifier 5 will be removed from the database. If you create a new user at "users", Firebase creates the identifier for you and assigns all the information you pass for the user. The paths follow the {{% a_blank "REST philosophy" "https://en.wikipedia.org/wiki/Representational_state_transfer" %}} where every entity (e.g. user, message, book, author) is associated with a URI, and HTTP methods are used to create, update, delete and get entities. In Firebase, the RESTful URI becomes a simple path, and the HTTP methods become Firebase's API.
```
Comment (rwieruch) [inline complete-firebase-authentication-react-tutorial.md]: "in Firebase's realtime API" -> "in Firebase's realtime database API"

File: complete-firebase-authentication-react-tutorial.md (hunk 7 above)
Comment (EmmanuelStalling) [reply complete-firebase-authentication-react-tutorial.md]: Should have been "with each other" or "with one another"

File: complete-firebase-authentication-react-tutorial.md (hunk 9 above)
Comment (EmmanuelStalling) [reply complete-firebase-authentication-react-tutorial.md]: You don't really ever need to use the word eventually in English writing.  You can say 'that will do something' and the eventually is assumed in context, as eventually just means 'in the future', and all things that haven't happened yet are eventual.  I changed this to "React components that will connect to the Firebase API"

File: complete-firebase-authentication-react-tutorial.md (hunk 10 above)
Comment (EmmanuelStalling) [reply complete-firebase-authentication-react-tutorial.md]: Unnecessary to tell the reader you'll explain something "in an elaborate way" (this phrase should be just the word "elaborately" anyway).  By the 700th line they know how elaborately you explain things.

File: complete-firebase-authentication-react-tutorial.md (hunk 11 above)
Comment (EmmanuelStalling) [reply complete-firebase-authentication-react-tutorial.md]: The dash is correct grammar here.

**Comment by rwieruch** [comment]
Perfect @EmmanuelStalling Got it merged 💯

**Pull request merged**